audioswitch/src/main/java/com/twilio/audioswitch/bluetooth/BluetoothHeadsetManager.kt

```diff
@@ -24,20 +26,44 @@ import com.twilio.audioswitch.android.BluetoothIntentProcessor
 import com.twilio.audioswitch.android.BluetoothIntentProcessorImpl
 import com.twilio.audioswitch.android.Logger
 import com.twilio.audioswitch.android.SystemClockWrapper
-import com.twilio.audioswitch.bluetooth.BluetoothHeadsetManager.HeadsetState.AudioActivated
-import com.twilio.audioswitch.bluetooth.BluetoothHeadsetManager.HeadsetState.AudioActivating
-import com.twilio.audioswitch.bluetooth.BluetoothHeadsetManager.HeadsetState.AudioActivationError
-import com.twilio.audioswitch.bluetooth.BluetoothHeadsetManager.HeadsetState.Connected
-import com.twilio.audioswitch.bluetooth.BluetoothHeadsetManager.HeadsetState.Disconnected
 
 private const val TAG = "BluetoothHeadsetManager"
 
-internal class BluetoothHeadsetManager
+internal interface BluetoothHeadsetManager {
+    fun start(headsetListener: BluetoothHeadsetConnectionListener)
+    fun stop()
+    fun activate()
+    fun deactivate()
+    fun hasActivationError(): Boolean
+    fun getHeadset(bluetoothHeadsetName: String?): AudioDevice.BluetoothHeadset?
+
+    companion object {
+        fun newInstance(
+            context: Context,
+            logger: Logger,
+            bluetoothAdapter: BluetoothAdapter?,
+            audioDeviceManager: AudioDeviceManager
+        ): BluetoothHeadsetManager? {
+            return bluetoothAdapter?.let { adapter ->
+                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
+                    BluetoothHeadsetManager_v31plus(context, logger, adapter, audioDeviceManager)
+                } else {
+                    BluetoothHeadsetManagerDefault(context, logger, adapter, audioDeviceManager)
+                }
+            } ?: run {
+                logger.d(TAG, "Bluetooth is not supported on this device")
+                null
+            }
+        }
+    }
+}
+
+internal open class BluetoothHeadsetManagerDefault
```

Comment:
An alternative approach would be to check for permissions in the original  `BluetoothHeadsetManager` similar to how you implemented `BluetoothHeadsetManager_v31plus`. The `checkPermissions()` function would look like

```
    private fun hasPermissions(): Boolean {
        return if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
            PERMISSION_GRANTED == context.checkSelfPermission(Manifest.permission.BLUETOOTH_CONNECT)
        } else {
            // This should always return true, but does cover cases where developers explicitly remove BLUETOOTH using manifest merging
            PERMISSION_GRANTED == context.checkCallingPermission(Manifest.permission.BLUETOOTH)
        }
    }
```

This approach removes the need to introduce two separate interfaces that are functionally the same and scope the changes to the checking permissions before interacting with BT APIs.

Reply:
@aaalaniz I don't understand this implementation because on pre Android 12, no user interaction is required for BLUETOOTH permission approval and the permission is always there because its part of AudioSwitch's AndroidManifest making it redundant... do you instead mean to wrap every public function in BluetoothHeadsetManagerDefault with ...
`private fun hasPermissions(): Boolean {
    return if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
            PERMISSION_GRANTED == context.checkSelfPermission(Manifest.permission.BLUETOOTH_CONNECT)
        } else {
            true
        }
   }`